documentation/projects/proposals/trust_and_safety/content_report_moderation/20230411-project_proposal_content_report_moderation.md

```diff
@@ -0,0 +1,338 @@
+  provider would require more work in the infrastructure repository to set up
+  configuration management and more places to track billing.
+- From an ethical perspective, essentially all the major computer vision APIs
+  have similar problems, so AWS Rekognition isn't an especially heinous choice.
+
+The implementation plan writer is welcome to explore other options they think
+might make sense in spite of those reasons listed above. However, Rekognition
+should be considered the "default" option unless another one provides a
+substantial benefit that invalidates those reasons.
+
+### Cache management (requirement 5)
+
+Our current caching strategy is to cache responses at the Cloudflare level for 1
+month based on query params. While this roughly works for our existing needs, it
+will not be workable for content moderation due to the inability to granularly
+invalidate cached pages via query params. Any Cloudflare cache features that
+would enable this like custom cache keys or cache tags, are all enterprise only
+features and are not accessible to Openverse.
+
+In order to invalidate responses that include a certain result, we need a
+reverse index of `result ID -> cache keys`. When a given result is marked as
+sensitive due to a report, we need to bust the cache keys under that result ID.
+Cache keys also need to be invalidated based on the response cache timeout.
+Redis does not support expiring set elements, but we can use a sorted set where
+the score is the unix timestamp of the expiration time. To expire cache keys in
+sorted sets that have expired cache keys, we can trigger a cache clean up
+function periodically during the regular request lifecycle[^workaround-kudos].
+
+[^workaround-kudos]:
+    Kudos to
+    [this old Google Groups discussion](https://web.archive.org/web/20211205091916/https://groups.google.com/g/redis-db/c/rXXMCLNkNSs)
+    for the sorted set workaround.
+
+Django's built in `cache_page` implementation does not work for our purposes out
+of the box as it naively caches responses without providing any utility for
+custom cache keys or to retrieve the cache key in downstream middleware. There
+are two potential approaches an implementation plan could take: we could
+implement our own caching middleware entirely that also maintains the reverse
+index; or we could monkey patch the `learn_cache_key` function used by the
+caching middlewares to attach the cache key on the request context to make it
+accessible to downstream middleware that manages the reverse index.
+
+### Moderation response times (requirement 6, optional)
+
+If we are able to identify volunteers to moderate Openverse's content report
+queue, we can work with them to establish expectations about response times. To
+have insight into the process, we'll need a reporting mechanism that tracks the
+number of resolved reports, average response time, and number of open reports on
+an ongoing basis.
```

Comment:
> we'll need a reporting mechanism that tracks the
number of resolved reports, average response time, and number of open reports on
an ongoing basis.

It isn't clear to me if this is optional or not, and if it is contingent on identifying volunteers or not? Would we want this even if Openverse maintainers were doing the moderation?

Reply:
@sarayourfriend just pinging you on this comment for visibility.

Reply:
I added further clarification that this section is contingent on finding the group of moderators.

> Would we want this even if Openverse maintainers were doing the moderation?

Right now I am working off the strong assumption that Openverse maintainers will not be the ones doing the moderation. As was shared with us from professional moderation folks, having non-professional moderation is fraught and likely to get us in trouble. My impression was that the people we spoke to were on board with that, but I didn't get an answer when I explicitly asked if they would help, so I'll follow up and ask again.

However, if Openverse maintainers _do_ end up doing the moderation, then sure, we'll want these things. Particularly if our team is being asked to supply OKRs and KPIs, these are easy and obvious ones to share and track.